File: Python/Invert_Tree.py
#Mirror Tree or Invert Tree

#-----------1st Way------------
class Node:
  def __init__(self, value):
    self.left = None
    self.right = None
    self.value = value
  def preorder(self):
    print(self.value, end=' ')
    if self.left: self.left.preorder()
    if self.right: self.right.preorder()

#function to swap left subtree with right subtree
def swap(node):
  if node is None:
  	return
  temp = node.left
  node.left = node.right
  node.right = temp

#Function to invert given binary Tree using preorder traversal
def invert(root):
	if root is  None:
		return
	swap(root)
	invert(root.left)
	invert(root.right)

root = Node('a') 
root.left = Node('b') 
root.right = Node('c') 
root.left.left = Node('d') 
root.left.right = Node('e') 
root.right.left = Node('f') 

root.preorder()
# a b d e c f 
print("\n")
invert(root)
root.preorder()

#-----------2nd Way------------
'''class Node:
  def __init__(self, value):
    self.left = None
    self.right = None
    self.value = value
  def preorder(self):
    print(self.value, end=' ')
    if self.left: self.left.preorder()
    if self.right: self.right.preorder()

#Function to swp and invert given binary Tree using preorder traversal
def invert(root):
	if root is  None:
		return
	temp = root
	invert(root.left)
	invert(root.right)
	temp = root.left
	root.left = root.right
	root.right = temp

root = Node('a') 
root.left = Node('b') 
root.right = Node('c') 
root.left.left = Node('d') 
root.left.right = Node('e')
root.right.left = Node('f') 

root.preorder()
# a b d e c f 
print("\n")
invert(root)
root.preorder()'''

File: C Language/Invert_Tree.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node* left;
    struct node* right;
};

struct node* newNode(int data)
{
    struct node* ptr = (struct node*)malloc(sizeof(struct node));
    ptr -> data = data;
    ptr -> left = NULL;
    ptr -> right = NULL;
    return ptr;
}

void preOrder(struct node* root)
{
    if(root == NULL)
    {
        return;
    }
    printf("%d ", root -> data);
    preOrder(root -> left);
    preOrder(root -> right);
}

struct node* invert(struct node* root)
{
    if(root == NULL)
    {
        return;
    }
    struct node* temp = root;
    invert(root -> left);
    invert(root -> right);
    
    temp = root -> left;
    root -> left = root -> right;
    root -> right = temp;
}

int main()
{
    struct node* root = newNode(1);
    root -> left = newNode(2);
    root -> right = newNode(3);
    root -> left -> left = newNode(4);
    root -> right -> right = newNode(5);
    
    printf("Before invert tree is:\n");
    preOrder(root);
    
    printf("\n\nAfter invert tree is:\n");
    invert(root);
    preOrder(root);
    
    return 0;
}